reputation/reputation.go

```diff
@@ -19,39 +25,165 @@ type ReputationModule struct {
 	ds      datastore.TxnDatastore
 	sources *source.SourceStore
 
+	mi *miner.MinerIndex
+
+	si *slashing.SlashingIndex
+
+	ai *ask.AskIndex
+
+	lockIndex sync.Mutex
+	mIndex    miner.Index
+	sIndex    slashing.Index
+	aIndex    ask.Index
+
+	lockScores sync.Mutex
+	rebuild    chan struct{}
+	scores     []MinerScore
+
 	ctx    context.Context
 	cancel context.CancelFunc
 }
 
-func New(ds datastore.TxnDatastore) *ReputationModule {
+type MinerScore struct {
+	Addr  string
+	Score int
+}
+
+func New(ds datastore.TxnDatastore, mi *miner.MinerIndex, si *slashing.SlashingIndex, ai *ask.AskIndex) *ReputationModule {
 	ctx, cancel := context.WithCancel(context.Background())
 	rm := &ReputationModule{
 		ds: ds,
+		mi: mi,
+		si: si,
+		ai: ai,
 
+		rebuild: make(chan struct{}, 1),
 		ctx:     ctx,
 		cancel:  cancel,
 		sources: source.NewStore(ds),
 	}
 
 	go rm.updateSources()
+	go rm.subscribeIndexes()
+	go rm.indexBuilder()
 
 	return rm
 }
 
+func (rm *ReputationModule) AddSource(id string, maddr ma.Multiaddr) error {
+	return rm.sources.Add(source.Source{Id: id, Maddr: maddr})
+}
+
+func (rm *ReputationModule) GetTopMiners(n int) ([]MinerScore, error) {
+	mr := make([]MinerScore, 0, n)
+	rm.lockScores.Lock()
+	for i := 0; i < n; i++ {
+		mr = append(mr, rm.scores[i])
+	}
+	rm.lockScores.Unlock()
+	return mr, nil
+}
+
 func (rm *ReputationModule) Close() error {
 	rm.cancel()
 	return nil
 }
 
-func (rm *ReputationModule) AddSource(id string, maddr ma.Multiaddr) error {
-	return rm.sources.Add(source.Source{Id: id, Maddr: maddr})
+func (rm *ReputationModule) subscribeIndexes() {
```

Comment:
This is a long running goroutine that listen for all indexes.
When some index say _Hey, I have new info_, it signals to recompute scores.